<a href="https://colab.research.google.com/github/LilianeEff/Data201/blob/main/Final_Project_Hints_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Import packages

In [ ]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import roc_auc_score, roc_curve, accuracy_score
from sklearn.metrics import precision_score, recall_score
from sklearn import preprocessing
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

## Read the csv file


In [ ]:
adult_df = pd.read_csv("adult.csv")

In [ ]:
adult_df.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18          ?  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                  ?    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

## The LabelEncode convers the categorical variables into numbers 0, 1, 2, ..
## An alternative method is to use pd.get_dummies()

In [ ]:
from sklearn.preprocessing import LabelEncoder
adult_df = adult_df.apply(LabelEncoder().fit_transform)


In [ ]:
adult_df.head()

age  workclass  fnlwgt  education  educational-num  marital-status  \
0    8          4   19329          1                6               4   
1   21          4    4212         11                8               2   
2   11          2   25340          7               11               2   
3   27          4   11201         15                9               2   
4    1          0    5411         15                9               4   

   occupation  relationship  race  gender  capital-gain  capital-loss  \
0           7             3     2       1             0             0   
1           5             0     4       1             0             0   
2          11             0     4       1             0             0   
3           7             0     2       1            98             0   
4           0             3     4       0             0             0   

   hours-per-week  native-country  income  
0              39              39       0  
1              49              39       0  
2              39              39       1  
3              39              39       1  
4              29              39       0

## Separate the features X and the label y

In [ ]:
#X = df.drop("income",axis=1)
#y = df.income

X = adult_df.iloc[:,0:-1]
y = adult_df.income

## Split the data into train and test

In [ ]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)

## Instantiate logistic regression and do the training

In [ ]:
LogReg = LogisticRegression()


#scaler = preprocessing.StandardScaler()
#X_train = scaler.fit_transform(X_train)
#X_test = scaler.transform(X_test)

LogReg.fit(X_train,y_train)


y_pred=LogReg.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Print the regression coefficients and the intercept

In [ ]:
print('Classes', LogReg.classes_)
print('Intercept',LogReg.intercept_)
print("Coefficients", LogReg.coef_ )

Classes [0 1]
Intercept [-0.02575355]
Coefficients [[ 3.76004716e-02 -7.26615842e-02 -1.86495192e-05 -5.44043321e-02
   1.95057633e-01 -1.89190295e-01 -8.20723328e-03 -2.05701793e-01
  -5.25892270e-02  2.94691630e-02  2.50265630e-02  6.94117984e-03
   1.01109626e-02 -6.50665812e-02]]


## Accuracy, precision, recall, and f1-score

In [ ]:
print("Accuracy", LogReg.score(X_test, y_test))
print(classification_report(y_test, LogReg.predict(X_test)))

Accuracy 0.7852755712062894
              precision    recall  f1-score   support

           0       0.81      0.93      0.87      9273
           1       0.60      0.33      0.42      2938

    accuracy                           0.79     12211
   macro avg       0.71      0.63      0.65     12211
weighted avg       0.76      0.79      0.76     12211



## Confusion Matrix

![i,age](https://assets-global.website-files.com/6266b596eef18c1931f938f9/644aea65cefe35380f198a5a_class_guide_cm08.png)

In [ ]:
conf_mat = confusion_matrix(y_test, y_pred)
conf_mat

array([[8620,  653],
       [1969,  969]])

False Positive Rate
<img src="https://wikimedia.org/api/rest_v1/media/math/render/svg/422d06161964ca90602ec8712cd211cb0d80da19">

From the confusion matrix:

$FPR = \frac{FP}{FP+TN} = \frac{1969}{1969+969}=0.67$

In [ ]:
df_male = adult_df[adult_df["gender"]==1]
X = df_male.drop("income",axis=1)
y = df_male.income
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)
LogReg = LogisticRegression()


scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

LogReg.fit(X_train,y_train)


y_pred=LogReg.predict(X_test)

In [ ]:
print('Classes', LogReg.classes_)
print('Intercept',LogReg.intercept_)
print("Coefficients", LogReg.coef_ )

Classes [0 1]
Intercept [-1.44682655]
Coefficients [[ 0.35379781 -0.05128913  0.06639697  0.07901792  0.87347914 -0.19779175
   0.0543393  -1.12328333  0.04571791  0.          0.65926592  0.30691887
   0.37281222  0.02064819]]


In [ ]:
print("Accuracy", LogReg.score(X_test, y_test))
print(classification_report(y_test, LogReg.predict(X_test)))

Accuracy 0.7932132794315815
              precision    recall  f1-score   support

           0       0.82      0.90      0.86      5661
           1       0.71      0.55      0.62      2502

    accuracy                           0.79      8163
   macro avg       0.77      0.72      0.74      8163
weighted avg       0.79      0.79      0.78      8163



In [ ]:
conf_mat = confusion_matrix(y_test, y_pred)
conf_mat

array([[5108,  553],
       [1135, 1367]])

From the confusion matrix:

$FPR = \frac{FP}{FP+TN} = \frac{1135}{1135+1367}=0.45$

In [ ]:
df_female = adult_df[adult_df["gender"]==0]
X = df_female.drop("income",axis=1)
y = df_female.income
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=0)
LogReg = LogisticRegression()


scaler = preprocessing.StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

LogReg.fit(X_train,y_train)


y_pred=LogReg.predict(X_test)

In [ ]:
print('Classes', LogReg.classes_)
print('Intercept',LogReg.intercept_)
print("Coefficients", LogReg.coef_ )

Classes [0 1]
Intercept [-3.07019866]
Coefficients [[ 0.45339203  0.06338293  0.04234591  0.06243343  1.00070441 -0.13551301
  -0.01314967  1.00091283  0.20281327  0.          0.59392373  0.24923453
   0.25524806  0.02440743]]


In [ ]:
print("Accuracy", LogReg.score(X_test, y_test))
print(classification_report(y_test, LogReg.predict(X_test)))

Accuracy 0.9197134387351779
              precision    recall  f1-score   support

           0       0.93      0.99      0.96      3593
           1       0.79      0.39      0.52       455

    accuracy                           0.92      4048
   macro avg       0.86      0.69      0.74      4048
weighted avg       0.91      0.92      0.91      4048



In [ ]:
conf_mat = confusion_matrix(y_test, y_pred)
conf_mat

array([[3547,   46],
       [ 279,  176]])

From the confusion matrix:

$FPR = \frac{FP}{FP+TN} = \frac{279}{279+176}=0.61$

## Compare the accuracy, precision, recall, and false positive rates of the females and males
